torch/fx/data_dependencies.py

```diff
@@ -0,0 +1,311 @@
+                    if output_node_name in input_partition.inputs:
+                        if input_partition_name not in output_partition.partition_dependents:
+                            output_partition.partition_dependents.append(input_partition_name)
+                            input_partition.partitions_dependent_on.append(output_partition_name)
+                        if output_node_name not in output_partition.outputs:
+                            # add node to list of output partition's list of output nodes
+                            output_partition.outputs.append(output_node_name)
+
+    # find partitions with no dependencies
+    root_partitions : List[str] = []
+    for partition_name, partition in partitions.items():
+        if not partition.partitions_dependent_on:
+            root_partitions.append(partition_name)
+
+    # check partitions for circular dependencies and create topological partition ordering
+    sorted_partitions : List[str] = []
+    while root_partitions:
+        root_partition = root_partitions.pop()
+        sorted_partitions.append(root_partition)
+        for dependent in partitions[root_partition].partition_dependents:
+            partitions[dependent].partitions_dependent_on.remove(root_partition)
+            if not partitions[dependent].partitions_dependent_on:
+                root_partitions.append(dependent)
+    if len(sorted_partitions) != len(partitions):
+        raise RuntimeError("cycle exists between partitions!")
+
+    # info needed to make outer module
+    outer_module_nodes : List[str] = []
+    submodule_hints = []
+
+    # add original placeholders to outer graph
+    for node_name, node in orig_nodes.items():
+        if node.op == 'placeholder':
+            outer_module_nodes.append(node_name)
+
+    # create submodules from partitions
+    for partition_name in sorted_partitions:
+        submodule_name = "submod_" + str(partition_name)
+        submodule_hints.append(submodule_name)
+        submodule_nodes = []
+
+        # track placeholders 
+        placeholder_hints = []
+        for node_name in partitions[partition_name].inputs:
+            node = orig_nodes[node_name]
+            if node.op not in ['placeholder', 'get_attr']: 
+                # node not originally placeholder, need to convert 
+                placeholder_hints.append(node_name)
+            submodule_nodes.append(node_name)
+
+        # grab rest of nodes from partition
+        submodule_nodes.extend(partitions[partition_name].nodes)
+
+        # create submodule itself
+        submodule, submodule_node = create_module_from_nodes(
+            module_name=submodule_name,
+            module_node_names=submodule_nodes,
+            module_outputs=partitions[partition_name].outputs,
+            orig_nodes=orig_nodes,
+            orig_root_module=root_m,
+            submodule_hints=[],
+            placeholder_hints=placeholder_hints,
+        )
```

Comment:
I think using a helper function for each of the submodules as well as the base graph may actually be making the code here more complicated. I took a stab at writing out what was in my head: https://gist.github.com/jamesr66a/089337e5e801c60a1ed4d59f0898931e

It's pretty long, so we can meet and discuss it offline, but in short, I think that constructing the submodules and base module are fundamentally different operations, and just coding them separately makes each of them simpler. In addition, I added some additional state to Partition (e.g. an environment map, a Graph, and Targets it refers to).

(note i think the behavior for get_attr in my version is different)